### DISCUSSION TITLE
clustering ?

### DISCUSSION THREAD

**Discussion by imkael** [jscad/OpenJSCAD.org#1257] [Q&A]
Hello and first of all congratulations for openjscad. I am currently developing an interface that uses the cli and often I find myself waiting by the calculation times and when I make a top on my computer, I realize that it is node that monopolizes a single cpu. Is there a way to execute commands like: openjscad example.jscad -o example.stl in a cluster node in order to take advantage of multiple CPUs and thus divide the calculation times? Thank you in advance for your feedback.

**Comment by z3dev** [comment]
@imkael WOW! That's a really good question.

Definitely, JavaScript is single threaded by design. It just simplifies everything, and makes implementations consistent. But... yeah. It doesn't scale on todays cpus.

Maybe, there's a solution. Let's see if someone has hit this before.

Why do you need more performance? Big projects? Huge scale shapes? User experience?

**Reply by imkael** [reply]
@z3dev  thanks for answer and all you have done and all you're doing for openjscad.
 it's essentialy for users experience or multi users at same time.
i devellope a little thing around openjscad for 3 years now and when i started it it was with 1.9 and everything looks for the moment like old time ( i'm not able to pass the project to 2.x for the moment) maybe i should...
executing time for some of my dessigns takes about 40 secondes. It's not a lot in absolute but i think web user lambda not able to wait so long time to have is design. i ve tryed to use pm2 to make cluster  but ojscad doesn't use them ...

**Reply by hrgdavor** [reply]
I am curious why you use cli lambda for web users ? Are you protecting your code behind the models ?

You could easily build models on user's browser and save some lambda dollars.

**Reply by imkael** [reply]
@hrgdavor hello and thanks for your answer effectivelly i need  webuser to indicate dimensions of pieces, after i compute results on server side before render. that's why i use cli and not  javascript on user side. and i don't know how to retreive dimensions from computer user.

**Reply by hrgdavor** [reply]
Since you are not using server to protect intellectual property, you should be able run to all that is jscad code in the browser.
It is also more cost effective and more scalable.

**Reply by z3dev** [reply]
@imkael you should be able to move the design from V1 to V2 fairly easy. V2 performance is far better that V1 as well. If you have any problems than just let us know, and we can provide some assistance.

by the way, later versions of NODEJS are faster as well.

**Reply by imkael** [reply]
t1u1 in is comment point exactly my problematic i'm worried about total time required to compute one job for one user. So i will try what @z3dev  suggested ( upgrading to V2 and nodejs) and be back if experience some trouble

**Reply by hrgdavor** [reply]
Still even if/when jscad will be able to parallelize lambda may not be better choice considering how many cores today computers/laptops/smartphones have.

I was wondering how many workers can a browser use, and when the number of users stops being beneficial
This test will calculate 22048M mod 97777
Benchmark Group 5: average: 517 ms, 16 Workers [8%]
Benchmark Group 4: average: 907 ms, 8 Workers [14%]
Benchmark Group 3: average: 1754 ms, 4 Workers [26%]
Benchmark Group 2: average: 3413 ms, 2 Workers [51%]
Benchmark Group 1: average: 6644 ms, 1 Worker [Standart Test, 100%]
http://pmav.eu/stuff/javascript-webworkers/

My laptop is a beastly gaming machine, so it actually got better all the way up to 16 threads (it has 16 logical cores)

bragging aside, this means web workers can be actually used for doing stuff faster.

Doing parallel single boolean op is more difficult, but if jscad would split modeling to generate a CSG tree then if there are multiple boolean operations some could be paralelised.
